examples/cross-browser.py

```diff
@@ -0,0 +1,152 @@
+        if "id" in message and message["id"] == command_id:
+            return message
+
+
+async def main():
+    # const browser = await puppeteer.launch();
+    # https://github.com/puppeteer/puppeteer/blob/4c3caaa3f99f0c31333a749ec50f56180507a374/examples/cross-browser.js#L29
+    websocket = await get_websocket()
+
+    # Not implemented:
+    # await browser.version());
+    # https://github.com/puppeteer/puppeteer/blob/4c3caaa3f99f0c31333a749ec50f56180507a374/examples/cross-browser.js#L32
+
+    # Puppeteer:
+    # const page = await browser.newPage();
+
+    # Part 1. Execute BiDi command and wait for result.
+    # ... await browser.newPage();
+    # https://github.com/puppeteer/puppeteer/blob/4c3caaa3f99f0c31333a749ec50f56180507a374/examples/cross-browser.js#L31
+    command_result = await run_and_wait_command({
+        "id": 1000,
+        "method": "PROTO.browsingContext.create",
+        "params": {}}, websocket)
+
+    # Part 2. Get the command result.
+    # const page = ...
+    # https://github.com/puppeteer/puppeteer/blob/4c3caaa3f99f0c31333a749ec50f56180507a374/examples/cross-browser.js#L31
+    # The `command_result` should be like this:
+    # {
+    #     "id": 1000,
+    #     "result": {
+    #         "context": "__SOME_CONTEXT_ID__",
+    #         "url": "",
+    #         "children": []
+    #     }
+    # }
+    context_id = command_result['result']['context']
+
+    # Puppeteer:
+    # await page.goto('https://news.ycombinator.com/');
+    # https://github.com/puppeteer/puppeteer/blob/4c3caaa3f99f0c31333a749ec50f56180507a374/examples/cross-browser.js#L34
+    await run_and_wait_command({
+        "id": 1001,
+        "method": "browsingContext.navigate",
+        "params": {
+            "url": "https://news.ycombinator.com/",
+            "context": context_id}}, websocket)
+
+    # Puppeteer:
+    # const resultsSelector = '.titlelink';
+    # https://github.com/puppeteer/puppeteer/blob/4c3caaa3f99f0c31333a749ec50f56180507a374/examples/cross-browser.js#L37
+    results_selector = {
+        "type": "string",
+        "value": ".titlelink"}
+
+    # Puppeteer:
+    # const links = await page.evaluate((resultsSelector) => {...}, resultsSelector);
+
+    # Part 1. Execute BiDi command.
+    # ... await page.evaluate((resultsSelector) => {...}, resultsSelector);
+    # https://github.com/puppeteer/puppeteer/blob/4c3caaa3f99f0c31333a749ec50f56180507a374/examples/cross-browser.js#L38
+    command_result = await run_and_wait_command({
+        "id": 1002,
+        "method": "PROTO.script.callFunction",
+        "params": {
+            "functionDeclaration": """(resultsSelector) => {
+                const anchors = Array.from(document.querySelectorAll(resultsSelector));
+                return anchors.map((anchor) => {
+                    const title = anchor.textContent.trim();
+                    return `${title} - ${anchor.href}`;
+                });
+            }""",
+            "args": [results_selector],
```

Comment:
It would be good if `run_and_wait_command` (or another helper) takes care of turning arguments into the right shape. Then, this could be simpler:

```suggestion
            "args": [".titlelink"],
```

Alternatively, ".titlelink" can be baked right into the function declaration since it's a constant. Maybe another example can have a more realistic use case for arguments?

Reply:
arguments above